feat: make the half-baked roles concept actually work

aheiska opened the pull request:
* **Please check if the PR fulfills these requirements**
- [X] The commit messages are descriptive
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [ ] An issue has been created for the pull requests. Some issues might require previous discussion.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This pr adds more features to roles. Previously roles could only refer to group or topic. With these changes role can be configured with arbitrary yaml key. See updated docs for more complete description.

Adds ability to configure multiple role files. Instead of one big, roles can be defined in multiple files.

Also adds error if a special_topic is defined more than once.

* **What is the current behavior?** (You can also link to an open issue here)

- Only group and topic can be used in role definition when granting acls
- Only one role file is supported

* **What is the new behavior (if this is a feature change)?**

- Extends previous behavior, old config works same as it did before.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

- Backwards compatible.

* **Other information**:

We julie for our kafka and found this is an only sane fork. There is also a PR offering these changes to original julie repo, but it seems to be very dead.

sverrehu commented:
There appears to be integration test failures. If you have Docker installed, you may run the tests locally using `mvn -Pacls integration-test`.

aheiska commented:
Fixed another one (JulieOpsIT). Another seems to spot an actual bug. I need dig bit deeper into this.

aheiska commented:
Ok, I think the tests are now fixed. Couple of notes

- Test files had `zookeepers:` elements in then. Which mapped to unused role causing an error so I removed them. This is undocumented and never functional feature (runtime list was always empty). 
- resources/roles.yaml had `permissionType: "READ"` and `"WRITE"` that are not valid values for `org.apache.kafka.common.acl.AclPermissionType`. This caused `TopologyBuilderAdminClient - java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Invalid ACL creation: Permission type is UNKNOWN`. No idea why this triggered now. Documentation had the same error.

So this grew a bit large. Some of it feels bug fixes to roles, which Pere I believe had never time to finish (the are indeed documented in future).

erikgb commented on docs/futures/define-custom-roles.rst:
```diff
@@ -125,4 +132,96 @@ Your topology file could look like this:
+            patternType: "LITERAL"
+            host: "*"
+            operation: "ALL"
+            permissionType: "ALLOW"
```
Would it make sense to make `permissionType: ALLOW` the default, or do we want to be very explicit here?

aheiska replied:
It would. And change is small. Added to PR.

runelan commented on src/test/java/com/purbon/kafka/topology/integration/AccessControlManagerIT.java:
```diff
@@ -526,6 +532,93 @@ private void verifyAclsOfSize(int size) throws ExecutionException, InterruptedEx
+    topology.setContext("integration-test");
+    topology.addOther("source", "testMirrorMakerAclsCreation");
+
+    var mm = new Other();
+    mm.setPrincipal("User:mm2");
+    mm.setGroup(Optional.of("testgroup"));
+    mm.setOtherField("statusTopic", "test-mirrormaker2-cluster-status");
+    mm.setOtherField("offsetTopic", "test-mirrormaker2-cluster-offsets");
+    mm.setOtherField("configTopic", "test-mirrormaker2-cluster-configs");
+    mm.setOtherField("targetPrefix", "test-mm.");
+    mm.setOtherField("offsetSyncTopic", "mm2-offset-syncs.test-mm.internal");
+    mm.setOtherField("checkpointsTopic", "test-mm.checkpoints.internal");
+
```
For the MirrorMaker test-case, should we add the heartbeat topic? Like the checkpoint, it is an optional topic for the connector, but I think we should cover that one as well

sverrehu commented:
I think this PR would be "cleaner" if the ZooKeeper removal was pulled out into a separate PR. What do you think, @aheiska?

aheiska commented:
> I think this PR would be "cleaner" if the ZooKeeper removal was pulled out into a separate PR. What do you think, @aheiska?

I agree. I'll try to make PR today.

aheiska commented:
> @aheiska, it seems like CI is failing, blocking us from cutting a release containing this: https://github.com/statnett/kafka-ops-julie/actions/runs/17484802829/job/49661666387?pr=426.
> 
> Can you look into it, please?
> 
> UPDATE: It seems like this was working yesterday, so there is a release with this: https://github.com/statnett/kafka-ops-julie/releases/tag/v5.2.0

Hmph, clean test and clean integration test on my local machine worked (from v5.2.0 and master). Error definitely refers to stuff I added though. 

Will try to reproduce.
